File: python/pca/Pcajb.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2020 Julian Bustamante <jbustamante35@gmail.com>
#
# Distributed under terms of the MIT license.

"""
My custom PCA class

"""

import numpy as np
from numpy import linalg as la
from magic import *

class Pcajb:
    def __init__(self, data, npc):
        self.Data        = np.array(data)
        self.NumberOfPCs = npc

    def meanSubtract(self):
        X = self.Data
        u = np.mean(X, axis=0)
        M = np.array(np.subtract(X,u))
        return M,u

    def covarMatrix(self):
        M,u = self.meanSubtract()
        npc = self.NumberOfPCs

        C = np.array(np.dot(np.transpose(M), M) / np.shape(M)[0])
        return C

    def eigens(self, neigs=None):
        if neigs is None:
            neigs = self.NumberOfPCs

        C   = self.covarMatrix()
        w,v = la.eig(C)
        return np.array(w[0:neigs]),np.array(v[:,0:neigs])

    def PCAScores(self, ndims=None, neigs=None):
        if ndims is None:
            sz    = np.shape(self.Data)
            ndims = np.arange(0, sz[0])

        if neigs is None:
            neigs = self.NumberOfPCs

        X   = self.Data[ndims,:]
        M,u = self.meanSubtract()
        w,v = self.eigens(neigs)
        return np.array(self.pcaProject(X,v,u,'sim2scr'))

    def SimData(self, ndims=None, neigs=None):
        if ndims is None:
            sz    = np.shape(self.Data)
            ndims = np.arange(0, sz[0])

        if neigs is None:
            neigs = self.NumberOfPCs

        M,u = self.meanSubtract()
        S   = self.PCAScores(ndims,neigs)
        w,v = self.eigens(neigs)
        return np.array(self.pcaProject(S,v,u,'scr2sim'))

    def VarExplained(self, pct=None, neigs=None):
        if pct is None:
            pct = 1.0

        if neigs is None:
            neigs = self.NumberOfPCs

        w,v = self.eigens(neigs)
        V   = np.array(np.cumsum(w / sum(w)))
        T   = V[V <= pct].size + 1
        return V,T

    def pcaProject(self,X,v,u,req):
        if req == 'scr2sim':
            return np.array(np.dot(X,np.transpose(v)) + u)

        elif req == 'sim2scr':
            return np.array(np.dot(X - u,v))

        else:
            return "Invalid projection"

File: r/pca/Pcajb.R
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
## ---------------------------
## Pcajb.R
## Purpose of script:
##
## Author: Julian Bustamante
##
## Date Created: 2020-03-05
##
## Copyright (c) Julian Bustamante, 2020
## Email: <jbustamante35@gmail.com>
##
## Distributed under terms of the MIT license.
##
## ---------------------------
##
## Notes:
##
##
## ---------------------------
library('pracma')

Pcajb <- setRefClass(
    # Class name
    "Pcajb",

    # Slots
    fields = list(
        Data = "matrix",
        npc  = "numeric"
    ),

    # Methods
    methods = list(
        meanSubtract = function() {
            x = Data
            u = colMeans(x)
            M = x - u
            Mu <- list(M,u)
            return(Mu)
        },

        covarMatrix = function() {
            mu = meanSubtract()
            m  = Reshape(unlist(mu[1]), size(Data,1), size(Data,2))
            c  = t(m) * m / size(m,1)
            return(c)
        },

        eigens = function(neigs=0) {
            if (neigs == 0) {
                 neigs = npc
            }

            c  = covarMatrix()
            mu = meanSubtract()
            wv = eigen(c)
            w  = Reshape(unlist(wv[1]), 1, size(Data,1))
            v  = Reshape(unlist(wv[2]), size(Data,1), size(Data,2))
            w  = w[1:neigs]
            v  = v[,1:neigs]
            wv = list(w,v)
            return(wv)
        },

        PCAScores = function(ndims=0,neigs=0) {
             if (ndims == 0) {
                 ndims = 1:size(Data,1)
             }

             if (neigs == 0) {
                 neigs = npc
             }

            x  = Data[ndims,]
            mu = meanSubtract()
            wv = eigens(neigs)

            u  = Reshape(unlist(mu[2]), 1, size(Data,2))
            v  = Reshape(unlist(wv[2]), size(Data,1), neigs)
            s  = pcaProject(x,v,u,'sim2scr')
            return(s)
        },

        SimData = function(ndims=0,neigs=0) {
             if (ndims == 0) {
                 ndims = 1:size(Data,1)
             }

             if (neigs == 0) {
                 neigs = npc
             }

            mu = meanSubtract()
            wv = eigens(neigs)
            s  = PCAScores(ndims,neigs)

            u  = Reshape(unlist(mu[2]), 1, size(Data,2))
            v  = Reshape(unlist(wv[2]), size(Data,1), neigs)
            xx = pcaProject(s,v,u,'scr2sim')
            return(xx)
        },

        VarExplained = function(pct=0,neigs=0) {
            if (pct == 0) {
                pca = 1.0
            }

            if (neigs == 0) {
                neigs = npc
            }

            wv = eigens(neigs)
            w  = Reshape(unlist(wv[1]), 1, neigs)
            V  = cumsum(w / sum(w))
            T  = length(V[V <= pct]) + 1
            VT = list(V,T)
            return(VT)
        },

        pcaProject = function(x,v,u,req) {
            if (req == 'scr2sim') {
                m = x %*% t(v)
                return(m + c(u))

            } else if (req == 'sim2scr') {
                m = x - c(u)
                return(m %*% v)

            } else {
                return("Invalid projection")
            }
        }
    )
)
